lib/grandpa/types.go

```diff
@@ -40,6 +43,35 @@ type State struct {
 	round  uint64   // voting round number
 }
 
+// NewState returns a new GRANDPA state
+func NewState(voters []*Voter, setID, round uint64) *State {
+	return &State{
+		voters: voters,
+		setID:  setID,
+		round:  round,
+	}
+}
+
+// pubkeyToVoter returns a Voter given a public key
+func (s *State) pubkeyToVoter(pk *ed25519.PublicKey) (*Voter, error) {
+	id := uint64(2^64) - 1
+
+	for i, v := range s.voters {
+		if bytes.Equal(pk.Encode(), v.key.Encode()) {
+			id = uint64(i)
```

Comment:
you could add a break statement, since after this expression there is no reason to continue the loop...